### ISSUE TITLE
gce: HTTPS health checks (feature request and implementation details)

### ISSUE THREAD

**Issue by dcarley** [hashicorp/terraform#2190]
We have [`google_compute_http_health_check`](https://www.terraform.io/docs/providers/google/r/compute_http_health_check.html) which configures GCE [`HttpHealthChecks`](https://cloud.google.com/compute/docs/reference/latest/httpHealthChecks) but nothing for [`HttpsHealthChecks`](https://cloud.google.com/compute/docs/load-balancing/http/https-api/httpsHealthChecks).

It seems that this could be implemented in one of two ways, each of which have a downside:

1. Create a new `google_compute_https_health_check` resource which would duplicate all of the code in `google_compute_http_health_check` and need to be maintained in parallel.
1. Add a `protocol` attribute to the existing `google_compute_http_health_check` resource which would involve a lot of conditionals or meta-programming to use the right type from the GCE library.

Do you have any preferences?

**Comment by dcarley** [comment]
@sparkprime @dhilton do you have any opinions?

**Comment by dhilton** [comment]
I'd :+1: an attribute - that way the code base would be more concise plus us consistent with how the AWS ELB provider makes use of an attribute: https://www.terraform.io/docs/providers/aws/r/elb.html

**Comment by sparkprime** [comment]
I would try to follow the official API as closely as possible.  What if future versions of the two resources diverge more than they do now?  If we add our own extra protocol field, then what if GCE adds its own protocol field to one of the resources.

I would not worry too much about maintenance of the Terraform code.  We can factor out common elements if necessary to avoid duplicating code / work.  I looked into doing this for instance <-> instance_template but decided it wouldn't be worth the hassle.  It may well be worth the hassle in this case.

**Comment by sparkprime** [comment]
By the way, I am very interesting in helping to implement HTTP load balancing support for GCE in Terraform.  Now that HTTPs is public, it seems like a great time to do it.

**Comment by dcarley** [comment]
Thanks.

Turns out this might be a moot point right now, because the `httpsHealthChecks` endpoint is only available from the beta API and thus not supported by the `compute/v1` library

:panda_face:

**Comment by sparkprime** [comment]
That's ok, it is really the resource that is beta, we can port everything over to the beta version of the API which should be backwards compatible.  We would mark the beta resource as beta in the documentation, as I did in my unmerged PR to merge the autoscaler / instance group manager stuff.

**Comment by dhilton** [comment]
Cool - so how do we make this happen?

**Comment by sparkprime** [comment]
Ok let's just switch everything to compute/beta, I'll do that now and run the full set of tests.  Then it should be just a case of refactoring the HTTP resource and using this to make an HTTPs resource.

**Comment by sparkprime** [comment]
Turns out this was not a good idea.  Changing the API to compute/beta means that explicit URLs given in the .tf with /v1/ in them will no-longer be valid, which is unnecessarily disruptive.  I'll instead add a new API endpoint called betaCompute specifically for making HTTPs resources.

**Comment by sparkprime** [comment]
Ok that is there in #2297

**Comment by dcarley** [comment]
@sparkprime I had a crack at this: https://github.com/hashicorp/terraform/compare/master...dcarley:f-https_health_check

However I ran into some problems, described in these two commits:

- https://github.com/hashicorp/terraform/commit/ff378140cf1638200696527c87d30e6f570bda24
- https://github.com/hashicorp/terraform/commit/b33fc3f8fe9330d8e9f7075487127e11568d46cd

Any thoughts?

**Comment by dcarley** [comment]
Welp, `compute/v0.beta` was removed and it hasn't been added to `compute/v1`, so I guess this is blocked.

**Comment by sparkprime** [comment]
I'll try to find out internally what's going on

**Comment by dcarley** [comment]
Thanks!

**Comment by sparkprime** [comment]
The v0.beta API is still there, but not discoverable.  That means a new build of the go client libraries didn't pick it up.  That is a bug and people are working on it, so it should re-appear soon.

**Comment by sparkprime** [comment]
For large values of "soon", apparently v0.beta is back now.  Sorry for the chaos :)

I recommend reverse-applying

https://github.com/svanharmelen/terraform/commit/e1d5af8ccf83935bcab3380b3b9822e7e0338d68

to your branch in order to get unblocked again

**Comment by sparkprime** [comment]
@lwander is this necessary for usage of https loadbalancer?

**Comment by lwander** [comment]
Unfortunately I didn't know we had this work open, so it's been reimplemented and merged into master already.

**Comment by sparkprime** [comment]
I can't find it -- this is the health check resource, not the proxy

**Comment by lwander** [comment]
Ah, right - it's not required since the URL Map resource corresponding to an HTTPS proxy can have a Backend Service checked by HTTP health checks, but it would be good to have. I'll add it in.

**Comment by lwander** [comment]
@sparkprime this has been merged - #3883

**Comment by sparkprime** [comment]
Thanks for the initial work, @dcarley
